feat(step): circular steps variant

ko2in opened the pull request:
## Description
This PR add the new variant `railing` to the steps which suppose to display the steps with circular railing horizontally or vertically.

I used the name `raliing` since we already have `rail` component which can conflict each other. If you have better name for this variant, please suggest.

## Testcase
https://jsfiddle.net/ko2in/n9dum7fg/

## Screenshot (if possible)

![fui-railing-step](https://github.com/fomantic/Fomantic-UI/assets/930315/256b56ec-043e-499c-bd08-96aee091b7e5)

## Closes
#2170

ko2in commented:
I had to fix stylelint errors and rebase to take out this commit, and done forced push. Is that be OK?

lubber-de commented:
Great! But why dont we just name it `circular` as we already have for image, icon, label and button?

lubber-de reviewed (commented):
The inner circle isnt centered, looks misplaced

![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/5d27cb74-847d-48d5-b024-9cf6d0fb7a58)

lubber-de reviewed (commented):
The `inverted` variant is not properly supported
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/08c5494a-c5c4-49f6-872b-d0ee7634fda8)

Using any other background than white would need to change the non completed icon background as well. I know you set white as background to cover the connection line (i had a similar issue for `connected` feed)
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/dcbf8251-f35c-4d6e-b452-f2ddd4a7141b)

lubber-de commented:
The connection line cover background could be avoided by

```css
.ui.railing.vertical.steps .step:not(:last-child) .content::before {
margin-top: 2.2em;
}
```
it seems....

I wonder why i had so many issues doing that for connected feed... perhaps more difficult markup, Will take a look for that again

ko2in commented:
> Great! But why dont we just name it `circular` as we already have for image, icon, label and button?

That sound also OK for me. I will update the PR with that name and other fixes.

lubber-de commented:
> I didn't implement for inverted yet. I don't know how the inverted should be at the moment.

Not much to be done i believe. I suggest to:
- use dedicated less variables (yes, even if those would have the same value) for the inverted variant
- increase specificity, so the normal inverted variant does not override the railing ones (would already fix the inner circle color from black to green)
- dont use white as background (perhaps no background at all/transparent in case my margin-top idea works for you would alrady fix that)
- use a light color for the connection lines
- use the default inverted green colors

jamessampford commented:
Nice! Would it be possible to have an option to have a title and/or description on the horizontal variant below the step circles?

lubber-de reviewed (commented):
The `ordered` variant is totally messed up unfortunately

![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/5f5fd361-230b-45f8-9ee6-5fc62928db84)

lubber-de reviewed (changes requested):
The size variants need adjustments at least for vertical variant

tiny steps
- check icon too low, but circle icon for uncompleted looks good now :D
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/b16ef861-45e9-45e8-80cc-1856142a289c)

massive steps
- both icons misplaced
- text for content title and description needs some padding (this is also the case for the normal step without your new variant, i just stumbled upon that) 
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/5c12a9e4-a97d-4f23-957a-99462f646e42)

lubber-de reviewed (commented):
The `link step` variant isn't supported

![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/c1135e84-1582-4367-8c46-9d024508af79)

lubber-de reviewed (commented):
The new variant seems to be always unstackable. This could be a design choice and perhaps supported later if a visually pleasing solution can be found

lubber-de reviewed (commented):
my final comment/suggestion: 😉 

`disabled step` does not have any visual effect (you removed that on purpose...why?)

We should support color names like we have for the `connected feed` variant.
Currently the green color is hardcoded (`@completedColor`). The green was supposed to work for the usual step variant where coloring would not make so much sense to me as it would only have effect on the check icon.

But as your new variant concentrates on the icon and connection line there is a possible need for a different color (maybe only for a single completed step) and screams for being colorizable via a color class name.

Support coloring as we have it elsewhere would support

-  `ui purple circular steps`
- `blue completed step`
- `teal step`
- The connection line should inherit the color as we have it in `connected feed`

ko2in commented:
> `disabled step` does not have any visual effect (you removed that on purpose...why?)

I don't see any difference style for disabled step like box item. Maybe, remove the small circle disc?

lubber-de commented:
> > `disabled step` does not have any visual effect (you removed that on purpose...why?)
> 
> I don't see any difference style for disabled step like box item. Maybe, remove the small circle disc?

A disabled variant almost always has
- a lower opacity value
- greyed out (if lower opacity doesn't make it visually clear already)
- pointer-events: none (already implemented)

ko2in commented:
Disabled state would OK with completed or active. The default state already looks similar to the disabled item as it's color and background is grayed out.

ko2in commented:
If this looks OK for you, then I'll apply also the opacity for the default state.
![rail-step-opacity](https://github.com/fomantic/Fomantic-UI/assets/930315/e56e9838-9cc3-4479-a93b-ebbee2c3ad19)

lubber-de commented:
the opacity can be applied to either normal state or completed state. But you should still see the default (non-disabled, non completed) state greyed out (if no color is given as suggested above) and opacity:1 (it gets more clear when text is used in the vertical variant

If you meant the same, then: yes 😄 

![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/a0742e3a-bc90-477f-b484-21a3c1680cfb)

lubber-de commented on src/definitions/elements/step.less:
```diff
@@ -251,49 +381,49 @@
 /* Mobile (Default) */
 @media only screen and (max-width: (@largestMobileScreen)) {
-    .ui.steps:not(.unstackable) {
+    .ui.steps:not(.unstackable, .railing) {
```
If you keep this css level 4 selector syntax, this has to wait until until 2.11 or even 3.0 as only browsers from January 2021 are supporting this syntax.
See https://github.com/fomantic/Fomantic-UI/discussions/2854 for voting results of new CSS level 4 feature inclusion poll

To avoid this use `:not(.unstackable):not(.circular)` instead

ko2in replied:
Thanks for your clarification. Will update PR very soon.

ko2in committed:
feat(step): introduce new variant circular step

The new variant circular step shows a series of steps with the
circular marker and connection line between them.

It supports the three layout styles: horizontal, vertical and
ordered.

It has the same states as box step: active, completed, disabled.
And it contains the variations: color, inverted, size.

Due to it's design nature, it doesn't have the variations such
as: stackble, unstackable, attached and evenly divided.

ko2in commented:
@lubber-de Here we go! I've finished the implementation with the other variantions and all your suggestions.

Due to the nature and design approaches, the circular step doesn't have the following variations: stackable, unstackable, attached and evenly divided.

I'm not sure what is going on with jsfiddle.net. It seems to be stopped working for me since yesterday. I cannot browse the website anymore.

So, I've created the examples and demonstrations in [CodeSandbox](https://codesandbox.io/s/fui-circular-step-hqc7k7?file=/index.html).

jamessampford commented:
Looking good 😊

The only bit I’m unsure about is the coloured line appearing after the active step, I think it should only go backwards to the previous step (so between completed and active)

Possibly should have an option not to change the header colour as well on the horizontal version (or possibly if by adding `.active` to the header it inherits the same colour?)

Any possibility of having header supported below the horizontal version?

Not sure if possible at all, but could even be pretty cool to have an optional pulsating active version (of the circle, could possibly have the line between active and last completed pulsating as well separately but certainly the active circle by itself)

lubber-de commented on src/themes/default/elements/step.variables:
```diff
@@ -101,6 +117,30 @@
+@circularStepBorder: @circularStepBorderWidth solid @circularStepColor;
+@circularStepRingSize: 2.2em;
+@circularStepIconSize: 0.95em;
+@circularStepIconDistance: calc((-@circularStepRingSize + @circularStepIconSize) - 0.35em);
```
reducing the distance to 0.3 instead of 0.35 made the inner circle look better centered to me (especially when compared to vertical variant)

0.35em:
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/da7aa8aa-2546-48f4-abf5-70fa4a22fc57)

0.3em:
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/5525252e-097b-4491-a99d-108cdaf29c79)

vertical (untouched)
![image](https://github.com/fomantic/Fomantic-UI/assets/18379884/b0e47955-ad8a-4f14-80b4-f14776bc96e9)

lubber-de commented on src/definitions/elements/step.less:
```diff
@@ -196,53 +324,107 @@
+                    & .content::before {
+                        content: "";
+                        position: absolute;
+                        border-left: @circularStepBorder;
```
An idea:
What if we make connection line optional as `ui connected circular steps`? This is how we already have it for `ui connected feed`
So only the border-left and each border-color setting would get an additional `connected` selector

What do you think?

ko2in replied:
Do you mean only for vertical step? Then, that's a good idea. 

But for the horizontal step, if the connection line isn't showing, then the UI looks awkward.
![horizontal-step-with-no-connection](https://github.com/fomantic/Fomantic-UI/assets/930315/d2686d28-37b5-4a13-89aa-00cedd7ecc0d)

In this case, we would need title and description like vertical step. And this would be a problem when the screen size is narrower.

lubber-de reviewed (changes requested):
👏🏼 This is already very awesome. Your example code is also a good candidate for later docs 😉 
Please see my comments again

ko2in commented on src/themes/default/elements/step.variables:
```diff
@@ -134,3 +174,10 @@
 @invertedActiveHoverBackground: #444;
 @invertedHoverBackground: #3f3f3f;
 @invertedDisabledBackground: #222;
+
+/* Railing Step */
+@railingStepActiveColor: @completedColor;
+@railingStepCompletedColor: @white;
+@railingStepCompletedBackground: @railingStepActiveColor;
+@railingStepCompletedIconHorizontalPosition: calc(@railingStepDiscPadding / 2);
+@railingStepCompletedIconVerticalPosition: calc((@railingStepDiscPadding * -1.5) + (@railingStepDiscPadding * 1.5));
```
Look like rebasing brought back these variables.

ko2in commented on src/definitions/elements/step.less:
```diff
@@ -164,6 +166,113 @@
+                    background: @circularStepCompletedBackground;
+                }
+                &.completed::after {
+                    left: calc(-@circularStepLineMargin + (calc((@circularStepLineMargin / 2) / 2)));
```
You are right. `em` always gives the same result regardless of the size. Let me refactor the `calc` that includes only `em` unit.

`left: calc(-@circularStepLineMargin + (calc((@circularStepLineMargin / 2) / 2)) + 1px);`

We cannot use the fixed value `1px` which can bring different result when the value of `@circularStepLineMargin` has changed. I thought that should be based on the half amount of `@circularStepBorderWidth`.

What do you think?

ko2in replied:
That additional pixel are now evaluated from `@circularStepBorderWidth`. The calc() with em units are kept, since they are evaluating with the variables which could be override by the user, where the default value of `2em / 2` and `4em / 2` in user customized variables would be different outcomes.

ko2in commented:
> The only bit I’m unsure about is the coloured line appearing after the active step, I think it should only go backwards to the previous step (so between completed and active)

@jamessampford Yeah! That would be better I guess. What do you think @lubber-de ?

ko2in committed:
Remove unused variables

ko2in committed:
Fix incorrect variation name

lubber-de commented:
Yes, agree

ko2in committed:
feat(step): change the rule about the color of rail line between step

Only the rail line after the completed step has particular color.
The other states use the default color for the rail line.

ko2in committed:
feat(step): refactor variables and calc() function

ko2in commented:
Now, the color rule for the rail line is changed as @jamessampford.

The line only after the completed step is now showing the particular color. The rest of the states would use the default color for the rail line.

**Jsfiddle:** https://jsfiddle.net/ko2in/86spora2/
**CodeSandbox:** https://codesandbox.io/s/fui-circular-step-hqc7k7?file=/index.html

jamessampford commented:
I’ve managed to add optional pulsating to active - https://codesandbox.io/s/fui-circular-step-forked-7tqhs6?file=/index.html - probably needs some tweaking … borrowed from transitions, though needs to target the `::before` - I did try `::after` as well, but think the former looks better on its own instead of the latter or both

ko2in commented:
> I’ve managed to add optional pulsating to active - https://codesandbox.io/s/fui-circular-step-forked-7tqhs6?file=/index.html - probably needs some tweaking … borrowed from transitions, though needs to target the `::before` - I did try `::after` as well, but think the former looks better on its own instead of the latter or both

@jamessampford You can submit your own PR later for any improvements. I would like to keep it as simple for now.

lubber-de reviewed (commented):
That's very awesome by now!
Just a little comment (see below) and a general question (not a request):

There are now a very lot of `:not(.circular)` selector enhancements.
Would it be easier (less code?) to instead override the circular setting (which would be needed for non circular only) instead directly. This is especially noticable for all the evenly divided variants (.x.steps)
I understand you did that because circular steps are styled differently, but perhaps it's worth some minutes to think about that again.

So instead of 
```css
.ui.steps:not(.circular) {
	display: inline-flex;
}
``` 

just override it for the ciscular variant like
```css
.ui.steps {
	display: inline-flex;
}
.ui.circular.steps {
	display: block;
}

```

ko2in commented on src/definitions/elements/step.less:
```diff
@@ -164,6 +166,111 @@
+                    left: @circularStepIconDistance;
+                }
+
+                &.link ::before,
+                &.link ::after {
+                    cursor: pointer;
+                }
```
That was a mistake. I'll remove this.

ko2in commented:
> That's very awesome by now! Just a little comment (see below) and a general question (not a request):
> 
> There are now a very lot of `:not(.circular)` selector enhancements. Would it be easier (less code?) to instead override the circular setting (which would be needed for non circular only) instead directly. This is especially noticable for all the evenly divided variants (.x.steps) I understand you did that because circular steps are styled differently, but perhaps it's worth some minutes to think about that again.
> 
> So instead of
> 
> ```css
> .ui.steps:not(.circular) {
> 	display: inline-flex;
> }
> ```
> 
> just override it for the ciscular variant like
> 
> ```css
> .ui.steps {
> 	display: inline-flex;
> }
> .ui.circular.steps {
> 	display: block;
> }
> ```

I'll try to review again and find out whether it's possible to avoid excluding with `:not(.circular)` and describe the specific style directly.

ko2in committed:
remove duplicate definitions

lubber-de commented:
I tried to remove the `:not(.circular)` for the evenly divided variants (.two.steps, .three.steps etc.) and it seems it works well with circular variant (because it only affects the width), so at least those can be removed i think, but you may check yourself, too

ivangomes commented:
Guys this project still good?

lubber-de committed:
fix(step): more compiler conditions

lubber-de committed:
feat(step): revert evenly divided, no effect as circular is flex 1

lubber-de committed:
feat(step): fix specificity and horizontal title

lubber-de committed:
feat(step): support horizontal content

lubber-de committed:
fix(step): linting fix

lubber-de committed:
Merge branch 'develop' into feat-railing-step

lubber-de commented:
I added support for horizontal content. 
https://jsfiddle.net/lubber/t1j7yxoz/37/
![image](https://github.com/user-attachments/assets/105fec89-dcd3-4cfd-92e9-aa6678b72036)

lubber-de commented:
Docs added by https://github.com/fomantic/Fomantic-UI-Docs/pull/538

ko2in commented:
@lubber-de Thanks your assistance in completing this task for me.
